### ISSUE TITLE
Non-zero exit status from HDPMI

### ISSUE THREAD

**Issue by jwt27** [Baron-von-Riedesel/HX#26]
Hi,

I noticed hdpmi32 always returns with a non-zero exit status, and it's causing
some problems.

I see it's used to signal if hdpmi was loaded via xms/vcpi/int15.  But I'm
wondering what that's used for.  Maybe some other part in the HX suite relies
on this?  It doesn't seem like very useful information for the user.

Reason I'm asking, is that djgpp's exe stub fails to load hdpmi if it returns
non-zero.  It does work somehow if you rename `hdpmi32.exe` to `cwsdpmi.exe`,
but I expect that is only if there is a second (real) `cwsdpmi.exe` in `PATH`
somewhere.

**Comment by Baron-von-Riedesel** [comment]
> I see it's used to signal if hdpmi was loaded via xms/vcpi/int15. But I'm wondering what that's used for. 

I can't remember. The decision what exit codes to set is now almost 30 years old.

**Comment by jwt27** [comment]
So, we don't know if anything will break if it's changed.

Not sure what the best course of action is.  I was considering sending a patch
to djgpp, but it would be a lot easier if hdpmi were made compatible with the
countless djgpp programs that already exist.

**Comment by jwt27** [comment]
Did some searching.  From what I can tell the exit status is only checked in `dpmildr.asm`, and it only checks if it's >= 4.  So to me it seems safe to make hdpmi always return zero on success, eg:

```patch
diff --git a/Src/HDPMI/HDPMI.INC b/Src/HDPMI/HDPMI.INC
index 1061b1f..bd36836 100644
--- a/Src/HDPMI/HDPMI.INC
+++ b/Src/HDPMI/HDPMI.INC
@@ -424,9 +424,7 @@ FH_XMS30	equ 80h
 
 ;------------------------ HDPMI exit codes
 
-;	00 - hdpmi installed in Int15/raw mode
-;	01 - hdpmi installed in XMS mode
-;	02 - hdpmi installed in VCPI mode
+;	00 - hdpmi installed
 ;	03 - DPMI host found, no VCPI host found, hdpmi not installed
 ;	04 - error: DOS memory low, cannot initialize
 ;	05 - error: cannot enable A20 in raw mode 
@@ -436,9 +434,7 @@ FH_XMS30	equ 80h
 ;	09 - error: CPU not 80386+
 ;	0A - error: invalid cmdline
 
-EXIT_HDPMI_IN_RAWMODE	equ 0
-EXIT_HDPMI_IN_XMSMODE	equ 1
-EXIT_HDPMI_IN_VCPIMODE	equ 2
+EXIT_HDPMI_OK		equ 0
 EXIT_DPMIHOST_RUNNING	equ 3
 EXIT_OUT_OF_DOSMEMORY	equ 4
 EXIT_CANNOT_ENABLE_A20	equ 5
diff --git a/Src/HDPMI/INIT.ASM b/Src/HDPMI/INIT.ASM
index 6bc4ef9..42b9a1b 100644
--- a/Src/HDPMI/INIT.ASM
+++ b/Src/HDPMI/INIT.ASM
@@ -961,14 +961,7 @@ endif
 	jc nodosmem3
 
 	@drprintf "initsvr_rm2: back in real mode, ss:sp=%X:%X",ss,sp
-	mov al,EXIT_HDPMI_IN_VCPIMODE
-	test byte ptr [fHost], FH_VCPI
-	jnz initsvr_ex
-	dec al
-	test byte ptr [fHost], FH_XMS
-	jnz initsvr_ex
-	dec al
-initsvr_ex::
+	mov al,EXIT_HDPMI_OK
 	@drprintf "initsvr_rm2: exits with ax=%X",ax
 	ret
 nodosmem3::
```

**Comment by jwt27** [comment]
By the way, I did send a patch to djgpp (but no response yet), since I think it's more correct to ignore the exit status, and it also saves a few bytes of stub space.  So if this were fixed on both sides then new djgpp programs would be compatible with old versions of hdpmi, and hdpmi would be compatible with existing djgpp programs.

**Comment by Baron-von-Riedesel** [comment]
And returning rc 3 (DPMI host already installed) is not a problem?

**Comment by jwt27** [comment]
Generally no, since the stub checks `int 2f/1687` before loading dpmi.  Only if the already-installed host is 16-bit, then it could be an issue.  From what I can tell hdpmi32 is already able to install itself over hdpmi16, and it will return 0 with the patch above.

Logically I think it makes sense to also return 0 if hdpmi detects that it's already installed (with the same version/mode), but for the djgpp stub this is not relevant.  As long as hdpmi is not able to install itself on top of another dpmi host, then rc 3 is a valid error condition.

For what it's worth, cwsdpmi just blindly installs over any existing dpmi host, even itself.  Don't know if you want to emulate that behavior.

**Comment by jwt27** [comment]
I see you're about to release v2.22.  Can we get the above patch in there too?
Seems there's not much interest in fixing this on the djgpp side.

**Comment by Baron-von-Riedesel** [comment]
> I see you're about to release v2.22.

Not really. Interrupt routing has been reworked in hdpmi, making a long test phase - and hence a prerelease - necessary.
